Subject: [PATCH] git-remote-mediawiki: support namespaces

By default git-remote-mediawiki only supports pages in the
'Main' (NS_MAIN = 0) namespace plus media files that can
be found in the 'File' (NS_FILE = 6) namespace if that's
enabled with the remote.<remote-name>.mediaimport config option.

When importing using 'by_page' (the default), all pages are
listed, but only in the NS_MAIN namespace, and then revisions
for each are imported.

When importing using 'by_rev' each revision is fetched but if
it refers to a page not in the pages list it will be skipped
even if all pages are requested.  Since the list of all pages
is generated only for the NS_MAIN namespace, pages in other
namespaces end up being skipped even though 'by_rev' was
requested via the remote.<remote-name>.fetchStrategy config option.

Fix this problem by allowing import of any page when fetching
using 'by_rev' and no user-specified pages or categories have
been given.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 contrib/mw-to-git/git-remote-mediawiki.perl | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/contrib/mw-to-git/git-remote-mediawiki.perl b/contrib/mw-to-git/git-remote-mediawiki.perl
index e4603da7..89cb879f 100755
--- a/contrib/mw-to-git/git-remote-mediawiki.perl
+++ b/contrib/mw-to-git/git-remote-mediawiki.perl
@@ -862,6 +862,8 @@ sub mw_import_revids {
 	my $n_actual = 0;
 	my $last_timestamp = 0; # Placeholer in case $rev->timestamp is undefined
 
+	my $user_defined = @tracked_pages || @tracked_categories;
+
 	foreach my $pagerevid (@{$revision_ids}) {
 	        # Count page even if we skip it, since we display
 		# $n/$total and $total includes skipped pages.
@@ -896,7 +898,7 @@ sub mw_import_revids {
 
 		my $page_title = $result_page->{title};
 
-		if (!exists($pages->{$page_title})) {
+		if ($user_defined && !exists($pages->{$page_title})) {
 			print {*STDERR} "${n}/", scalar(@{$revision_ids}),
 				": Skipping revision #$rev->{revid} of ${page_title}\n";
 			next;
---
